### ISSUE TITLE
Critisms of main example provided.

### ISSUE THREAD

**Issue by Skarlett** [alexcrichton/tokio-process#53]
```rust
extern crate futures;
extern crate tokio;
extern crate tokio_process;
extern crate tokio_io;

use std::io;
use std::process::{Command, Stdio};

use futures::{Future, Stream};
use tokio_process::{CommandExt, Child};

fn print_lines(mut cat: Child) -> Box<Future<Item = (), Error = ()> + Send + 'static> {
    let stdout = cat.stdout().take().unwrap();
    let reader = io::BufReader::new(stdout);
    let lines = tokio_io::io::lines(reader);
    let cycle = lines.for_each(|l| { /* Pain point */
        println!("Line: {}", l);
        Ok(())
    });

    let future = cycle.join(cat) /* Pain point */
        .map(|_| ())
        .map_err(|e| panic!("{}", e));

    Box::new(future)
}

fn main() {
    let mut cmd = Command::new("cat");
    cmd.stdout(Stdio::piped());

    let future = print_lines(cmd.spawn_async().expect("failed to spawn command"));
    tokio::run(future);
}
```

There are a few pain points in the example provided, and I'll explain why they're rather difficult to work with. Seeing how we're using the highest level abstractions in the library, this is due for some scrutiny.

First and foremost, using `for_each` in an async is rather uncomfortable, and usually not a problem. But in this example, it actually tends to be because we usually want the return to continue pushing the data through the iterator, so we can build ontop of the future. The reason the example falls vastly short is because it's forcing us to use `()` as our return. meaning all actions must occur now, and escaping the data out of the iterator is nearly impossible. The caveat to doing is this, is most people will immediately just change their return future to match the type required.

```rust
fn print_lines(mut cat: Child) -> Box<Future<Item = SocketAddr, Error = AddrParseError> + Send + 'static> {
// snip
let cycle = lines.for_each(|l|
    l.parse::<SocketAddr>()
});
```
Only to fall victim to 3 new problems they've just created. First off - `for_each` now complains because it can't handle the error. `expected struct "std::net::AddrParseError", found struct "std::io::Error"`

Second problem, suddenly the `join` method isn't a method. This is most likely because `for_each` expects `future::FutureResult<_, std::io::Error>`

So you might think you're clever and just try to force a type correction in the loop by using `future::ok` with a specified tailfish.

e.g
```
fn print_lines(mut cat: Child) -> Box<Future<Item=SocketAddr, Error = Box<std::error::Error>> + Send + 'static>
// snip
let cycle = lines.for_each(|l| {
    future::ok::<SocketAddr, Box<std::error::Error>>(l.parse::<SocketAddr>().unwrap())
});
```
Yet this causes a totally new problem. A type mismatch (Wasn't that what we were just trying to fix?!)
`<futures::future::result_::FutureResult<std::net::SocketAddr, std::boxed::Box<dyn std::error::Error>> as futures::future::IntoFuture>::Error == std::io::Error`. 
[Note] expected struct `std::boxed::Box`, found struct `std::io::Error`

Even moving past all these type errors and tedious type mismatches, we can eventually end up in a point where we want to process the `for_each` routine with another `map` or `fold` - Well then we lost the function `join` later on in the code. `let future = cycle.join(cat)...` - ending up with something that looks like
 `[E0599] no method named "join" found for type futures::stream::for_each::ForEach<tokio_io::lines::Lines<std::io::BufReader<tokio_process::ChildStdout>>, [closure@src/genscripts.rs:74:32: 76:6], futures::future::result_::FutureResult<std::net::SocketAddr, std::boxed::Box<dyn std::error::Error>>> in the current scope`
 

In conclusion - you should provide an example that allows us to return the data inside the iterator for further processing, and attempt to help mitigate type errors conflicting with such functions.

**Comment by ipetkov** [comment]
Hey @Skarlett thanks for the detailed report!

You're right, the example isn't the best fit for plugging code into, I suppose it was meant to give a quick and dirty example of how to invoke the `tokio-process` APIs (IIRC it was probably also written back when were were on `tokio-core` so maybe it doesn't use the latest tokio APIs).

I think the first painpoint stems from the fact that `tokio::run` requires that the future's Item and Error are both `()`. One possibility is to change the return type's Error to be `failure::Error` (which allows for more ergonomic conversion from other errors via `?`), and modifying the actual `tokio::run` call to handle/panic on errors.

I think the second painpoint also seems to stem from the fact that `tokio::run` needs the Item type to be `()` as well, and for convenience the `print_lines` function simply prints things out for convenience and throws them away. We also need to be careful not to drop the spawned cat command or else the process will get killed. Maybe if we pass it directly to tokio via `tokio::spawn` we could change `print_lines` to return a stream of lines and allow the caller to consume the lines in another way.

What do you think?

**Comment by Skarlett** [comment]
I will start by saying I'm no expert at asynchronous programming, and would consider myself a noob in the specifics of the field. So no opinions here should be taken as stone. With that being said - I can only speak as a user of the library.

I will say this is hard problem, and by no means I have absolute answers, but I have a few ideas.

Before we get into what **I believe** is the technical details needed to accomplish this, I would like to lay out some basic acknowledgment for what it is that we're trying to achieve.
  - Allowing the future to return something other than `()` so it can be polled and collected upon later.
  (Edit this if anyone extends the issue)  

The ideal solution I would imagine is extending `map` and `fold` to work with `join`, or providing an abstraction that will chain the two futures, one processing into the other. It appears this was attempted, but might be really hard to get right (Resulting in what we currently have), especially in edge conditions where errors might need to be handled.

I suppose running `tokio::spawn` inside of `for_each` would be unnecessary simply because it will queue each line into the thread pool as each its own future - The reason I believe this isn't a good long term fix is because then you're filling the queue with each future needing to be polled for each line. Though I do not know how badly this would be for performance. This is the behavior we desire, but it populates a large portion of the queue with routines that should just be a future that resolves each new line for processing.

Using channels might be the way to go with this using the existing `for_each` - the problem with that is getting that async channel to resolve (We can't return a future), otherwise you'll have to use non-async channels. (That means we have an active thread wasting CPU cycles)

The other solution is to fore-go the entirety of reading a process line by line, and simply waiting for the process to end. (This is actually what I ended up doing in my code base) But in that case, you have to assume the process will end.

So to conclude, I believe we can ignore the error problems if we can get `map` and `fold` to be compatible with `join` because error handling can be implemented in the logic of the future. Though as I stated before, this may not be possible, for the reasons of `tokio::run`'s return requirement. Regardless of the complications, the desired feature is to have an async line-by-line read. I don't have enough experience to say that filling the queue with `tokio::spawn` is a bad idea, I just don't its the most elegant way (But it may be the only way). 

At the end of the day, these are just my thoughts, I hope other people can join in the conversation - give some ideas, work arounds, and some more expertise in the field than I can.

**Comment by ipetkov** [comment]
Hey @Skarlett, I've rewritten the example to be a bit more flexible and more easily extendible given your examples.

Feel free to reopen if there are additional pain points which I might have missed!
